enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,722 @@
+configuration profiles, focusing only on tests likeliest to fail based on the 
+configuration change, in order to reduce overall execution time.
+
+`openshift-tests` will plan an efficient testing strategy based on the 
+extension binaries it discovers, the configuration profiles they support,
+and the tests that need to run in each configuration. It will:
+1. Discover extension binaries.
+2. Discover extension binary metadata (`info` verb), including configuration profiles.
+3. Discover tests that should be run for each configuration profile (`list` verb with `--config <profile>`, for each profile).
+4. Plan test execution by grouping configuration profiles into buckets.
+
+After `openshift-tests` plans test orchestration, it will proceed to execute tests in
+each configuration bucket. For a given bucket, `openshift-tests`
+will:
+1. Iterate through all involved components and configure them with non-conflicting configuration profiles (`config` verb).
+2. Await a high-level cluster steady state (cluster operators available, machines config up-to-date) if any configuration profile requires it.
+3. Run tests within the bucket with non-conflicting, test-level parallelization.
+
+Between buckets, the component may be asked to set its required `default` profile in order
+to return to its install-time configuration.
+
+#### Update - Metadata Validation
+Component owners will be responsible for implementing the extension interface. To prevent common mistakes
+and ensure conformance with the evolution of their implementation, `make` (or similar build system)
+must run `<extension binary> [component parameters] update [--basedir <basedir>]` after the extension binary is built.
+
+The `update` verb will create or update files under `hack/.openshift-tests-exension/product/type/component/*`, by default
+(basedir defaults to `hack/.openshift-tests-extension`).
+If an incompatible change is introduced from the prior invocation of `update` (e.g. changing 
+a test name without preserving the original), `update` will raise an error 
+which the component owner must correct before committing their change in git.
+
+The content of the files stored under `.openshift-tests-extension` is subject to change and wholly
+defined by the `github.com/openshift-eng/openshift-tests-extension` module. Individual component
+owners should not modify files under this path except through invocations of `update`.
+
+#### Test Result Aggregation
+
+`openshift-tests` will store test results in the CI artifacts repository in a JSONL file.
+The output will be similar to that of the `run-test` output, but each test result 
+will be expanded to include:
+- Component & subcomponent
+- Environment
+- Testing Context
+
+The goal here is to allow a single file to contain comprehensive information necessary for
+a system like Component Readiness to analyze it without, for example, needing to draw information
+from prowjob names.

```

Comment:
We'll either need to produce junit xml for visualization in spyglass, or have spyglass support the new format.  I almost wonder if we shouldn't break this off into its own enhancement and come up with a modern spec for test results (https://xkcd.com/927/).  Or see if an existing standard fits our needs.

Reply:
Since the output of `run-tests` is a super-set of the existing junit XML (and origin aggregates and writes both files), we should be able to populate the old and new formats for as long as necessary for a smooth transition. TRT can migrate to the richer metadata source when they have time. Migrating spyglass is optional until we have a use case for the additional metadata.